File: src/main/java/voidchess/figures/King.java
package voidchess.figures;

import voidchess.board.BasicChessGameInterface;
import voidchess.board.SimpleChessBoardInterface;
import voidchess.helper.CheckSearch;
import voidchess.helper.Move;
import voidchess.helper.Position;
import voidchess.image.ImageType;

import java.util.List;

/**
 * @author stephan
 */
public class King extends RochadeFigure {
    private boolean didRochade;

    public King(boolean isWhite, Position startPosition) {
        super(isWhite, startPosition, FigureType.KING);
        didRochade = false;
    }

    public King(boolean isWhite, Position startPosition, int stepsTaken, boolean didRochade) {
        super(isWhite, startPosition, stepsTaken, FigureType.KING);
        this.didRochade = didRochade;
    }

    public boolean isReachable(Position to, BasicChessGameInterface game) {
        int horizontal_difference = Math.abs(getPosition().getRow() - to.getRow());
        int vertical_difference = Math.abs(getPosition().getColumn() - to.getColumn());
        if (horizontal_difference <= 1 && vertical_difference <= 1) {
            if (game.isFreeArea(to) || hasDifferentColor(game.getFigure(to))) return true;
        }
        if (((getPosition().getRow() == 0 && isWhite()) || (getPosition().getRow() == 7 && !isWhite())) && horizontal_difference == 0) {
            if (isShortRochadeReachable(to, game)) return true;
            if (isLongRochadeReachable(to, game)) return true;
        }
        return false;
    }

    private boolean isShortRochadeReachable(Position to, BasicChessGameInterface game) {
        if (canParticipateInRochade() &&
                !game.isFreeArea(to) &&
                to.getColumn() > getPosition().getColumn()) {
            int ground_row = isWhite() ? 0 : 7;
            Figure toFigure = game.getFigure(to);

            if (toFigure.canParticipateInRochade() && toFigure.isWhite() == isWhite()) {
                if (getPosition().getColumn() == 6) {
                    if (!game.isFreeArea(Position.Companion.get(ground_row, 5))) {
                        return false;
                    }
                } else {
                    //Die Felder bis zur g-Spalte müssen bis auf den Turm leer sein
                    for (int column = getPosition().getColumn() + 1; column < 7; column++) {
                        Position midlePosition = Position.Companion.get(ground_row, column);
                        if (!game.isFreeArea(midlePosition) &&
                                !game.getFigure(midlePosition).canParticipateInRochade()) {
                            return false;
                        }
                    }
                }
                return true;
            }
        }
        return false;
    }

    private boolean isLongRochadeReachable(Position to, BasicChessGameInterface game) {
        if (canParticipateInRochade() &&
                !game.isFreeArea(to) &&
                to.getColumn() < getPosition().getColumn()) {
            int ground_row = isWhite() ? 0 : 7;
            Figure toFigure = game.getFigure(to);

            if (toFigure.canParticipateInRochade() && toFigure.isWhite() == isWhite()) {
                //kommt der König auf die c-Linie?
                if (getPosition().getColumn() == 1) {        //auf der a-Linie kann der König nicht stehen, da dort Turm sein muß
                    if (!game.isFreeArea(Position.Companion.get(ground_row, 2))) {
                        return false;
                    }
                } else if (getPosition().getColumn() > 2) {
                    //Die Felder bis zur c-Spalte müssen bis auf den Turm leer sein
                    for (int column = getPosition().getColumn() - 1; column >= 2; column--) {
                        Position midlePosition = Position.Companion.get(ground_row, column);
                        if (!game.isFreeArea(midlePosition) &&
                                !game.getFigure(midlePosition).canParticipateInRochade()) {
                            return false;
                        }
                    }
                }
                //kommt der Turm auf die d-Linie?
                if (to.getColumn() != 3) {
                    final int step = CheckSearch.signum(3 - to.getColumn());
                    Position midlePosition;
                    Figure midleFigure;
                    for (int column = to.getColumn() + step; column != 3; column += step) {
                        midlePosition = Position.Companion.get(ground_row, column);
                        if (!game.isFreeArea(midlePosition)) {
                            midleFigure = game.getFigure(midlePosition);
                            if (!midleFigure.canParticipateInRochade() ||
                                    (midleFigure.isRock())) {
                                return false;
                            }
                        }
                    }
                    midlePosition = Position.Companion.get(ground_row, 3);
                    if (!game.isFreeArea(midlePosition)) {
                        midleFigure = game.getFigure(midlePosition);
                        if (!midleFigure.canParticipateInRochade() ||
                                midleFigure.isRock()) {
                            return false;
                        }
                    }
                }
                return true;
            }
        }
        return false;
    }

    public boolean isPassiveBound(Position to, SimpleChessBoardInterface game) {
        Position realTo = to;
        if (!game.isFreeArea(to) && game.getFigure(to).canParticipateInRochade()) {
            final int column = (to.getColumn() - getPosition().getColumn()) > 0 ? 6 : 2;
            realTo = Position.Companion.get(to.getRow(), column);
            if (CheckSearch.isCheck(game, getPosition())) return true;
            if (isKingAtCheckInbetweenRochade(getPosition(), realTo, game)) return true;
        }
        return isKingCheckAt(realTo, game);
    }

    private boolean isKingAtCheckInbetweenRochade(
            Position from,
            Position to,
            SimpleChessBoardInterface game
    ) {
        assert from.getRow() == to.getRow();

        final int step = CheckSearch.signum(to.getColumn() - from.getColumn());
        for (int column = from.getColumn() + step; column != to.getColumn(); column += step) {
            if (isKingCheckAt(Position.Companion.get(from.getRow(), column), game)) return true;
        }
        return false;
    }

    private boolean isKingCheckAt(Position to, SimpleChessBoardInterface game) {
        game.setFigure(getPosition(), null);
        Figure hitFigure = game.getFigure(to);
        game.setFigure(to, this);
        boolean isCheck = CheckSearch.isCheck(game, to);
        game.setFigure(to, hitFigure);
        game.setFigure(getPosition(), this);
        return isCheck;
    }

    public void getReachableMoves(BasicChessGameInterface game, List<Move> result) {
        int minRow = Math.max(getPosition().getRow() - 1, 0);
        int minColumn = Math.max(getPosition().getColumn() - 1, 0);
        int maxRow = Math.min(getPosition().getRow() + 1, 7);
        int maxColumn = Math.min(getPosition().getColumn() + 1, 7);

        for (int row = minRow; row <= maxRow; row++) {
            for (int column = minColumn; column <= maxColumn; column++) {
                Position checkPosition = Position.Companion.get(row, column);
                if (!checkPosition.equalsPosition(getPosition()) &&
                        (game.isFreeArea(checkPosition) ||
                                game.getFigure(checkPosition).isWhite() != isWhite())
                ) {
                    result.add(Move.Companion.get(getPosition(), checkPosition));
                }
            }
        }

        if (canParticipateInRochade()) {
            for (int column = getPosition().getColumn() + 1; column < 8; column++) {
                Position pos = Position.Companion.get(getPosition().getRow(), column);
                if (!game.isFreeArea(pos)) {
                    if (game.getFigure(pos).canParticipateInRochade() &&
                            isReachable(pos, game)) {

                        result.add(Move.Companion.get(getPosition(), pos));
                    }
                    break;
                }
            }
            for (int column = getPosition().getColumn() - 1; column >= 0; column--) {
                Position pos = Position.Companion.get(getPosition().getRow(), column);
                if (!game.isFreeArea(pos)) {
                    if (game.getFigure(pos).canParticipateInRochade() &&
                            isReachable(pos, game)) {

                        result.add(Move.Companion.get(getPosition(), pos));
                    }
                    break;
                }
            }
        }
    }

    public boolean isSelectable(SimpleChessBoardInterface game) {
        int minRow = Math.max(getPosition().getRow() - 1, 0);
        int minColumn = Math.max(getPosition().getColumn() - 1, 0);
        int maxRow = Math.min(getPosition().getRow() + 1, 7);
        int maxColumn = Math.min(getPosition().getColumn() + 1, 7);

        for (int row = minRow; row <= maxRow; row++) {
            for (int column = minColumn; column <= maxColumn; column++) {
                Position checkPosition = Position.Companion.get(row, column);
                if (isMovable(checkPosition, game)) {
                    return true;
                }
            }
        }

        if (getPosition().getColumn() + 2 < 8) {
            Position shortRochade = Position.Companion.get(getPosition().getRow(), getPosition().getColumn() + 2);
            if (isMovable(shortRochade, game)) {
                return true;
            }
        }

        if (getPosition().getColumn() - 2 >= 0) {
            Position longRochade = Position.Companion.get(getPosition().getRow(), getPosition().getColumn() - 2);
            if (isMovable(longRochade, game)) {
                return true;
            }
        }

        return false;
    }

    public int countReachableMoves(BasicChessGameInterface game) {
        int count = 0;
        int minRow = Math.max(getPosition().getRow() - 1, 0);
        int minColumn = Math.max(getPosition().getColumn() - 1, 0);
        int maxRow = Math.min(getPosition().getRow() + 1, 7);
        int maxColumn = Math.min(getPosition().getColumn() + 1, 7);

        for (int row = minRow; row <= maxRow; row++) {
            for (int column = minColumn; column <= maxColumn; column++) {
                Position checkPosition = Position.Companion.get(row, column);
                if (isReachable(checkPosition, game)) {
                    count++;
                }
            }
        }

        if (getPosition().getColumn() + 2 < 8) {
            Position shortRochade = Position.Companion.get(getPosition().getRow(), getPosition().getColumn() + 2);
            if (isReachable(shortRochade, game)) {
                count++;
            }
        }

        if (getPosition().getColumn() - 2 >= 0) {
            Position longRochade = Position.Companion.get(getPosition().getRow(), getPosition().getColumn() - 2);
            if (isReachable(longRochade, game)) {
                count++;
            }
        }

        return count;
    }

    public boolean didRochade() {
        return didRochade;
    }

    public void performRochade() {
        didRochade = true;
    }

    public void undoMove(Position oldPosition) {
        super.undoMove(oldPosition);

        if (stepsTaken == 0) {
            didRochade = false;
        }
    }

    public String toString() {
        String parentString = super.toString();
        StringBuilder buffer = new StringBuilder(parentString.length() + 5);

        buffer.append(parentString);
        if (didRochade) buffer.append("-true");

        return buffer.toString();
    }
}

File: src/main/kotlin/voidchess/figures/King.kt
package voidchess.figures

import voidchess.board.BasicChessGameInterface
import voidchess.board.SimpleChessBoardInterface
import voidchess.board.check.AttackLines
import voidchess.board.check.BoundLine
import voidchess.board.check.CheckLine
import voidchess.board.check.CheckSearch
import voidchess.board.move.Direction
import voidchess.board.move.Move
import voidchess.board.move.Position
import java.lang.IllegalArgumentException
import java.lang.UnsupportedOperationException
import kotlin.math.sign


class King : CastlingFigure {
    private var didCastling: Boolean = false
    private val groundRow = if (isWhite) 0 else 7

    constructor(isWhite: Boolean, startPosition: Position) : super(isWhite, startPosition, FigureType.KING) {
        didCastling = false
    }

    constructor(isWhite: Boolean, startPosition: Position, stepsTaken: Int, didCastling: Boolean) : super(isWhite, startPosition, stepsTaken, FigureType.KING) {
        this.didCastling = didCastling
    }

    override fun isReachable(toPos: Position, game: BasicChessGameInterface): Boolean {
        val horizontalDifference = Math.abs(position.row - toPos.row)
        val verticalDifference = Math.abs(position.column - toPos.column)
        if (horizontalDifference <= 1 && verticalDifference <= 1) {
            val figure = game.getFigureOrNull(toPos)
            if (figure==null || hasDifferentColor(figure)) return true
        }
        if (isShortCastlingReachable(toPos, game)) return true
        return isLongCastlingReachable(toPos, game)
    }

    private fun isShortCastlingReachable(to: Position, game: BasicChessGameInterface): Boolean {
        val toFigure = game.getFigureOrNull(to)
        if (canCastle() &&
                toFigure!=null &&
                to.column > position.column) {
            if (toFigure.canCastle() && toFigure.isWhite == isWhite) {
                if (position.column == 6) {
                    if (!game.isFreeArea(Position[groundRow, 5])) {
                        return false
                    }
                } else {
                    //Die Felder bis zur g-Spalte müssen bis auf den Turm leer sein
                    for (column in position.column + 1..6) {
                        val middlePosition = Position[groundRow, column]
                        val middleFigure = game.getFigureOrNull(middlePosition)
                        if (middleFigure!=null && !middleFigure.canCastle()) {
                            return false
                        }
                    }
                }
                return true
            }
        }
        return false
    }

    private fun isLongCastlingReachable(to: Position, game: BasicChessGameInterface): Boolean {
        val toFigure = game.getFigureOrNull(to)
        if (canCastle() &&
                toFigure!=null &&
                to.column < position.column) {


            if (toFigure.canCastle() && toFigure.isWhite == isWhite) {
                //kommt der König auf die c-Linie?
                if (position.column == 1) {        //auf der a-Linie kann der König nicht stehen, da dort Turm sein muß
                    if (!game.isFreeArea(Position[groundRow, 2])) {
                        return false
                    }
                } else if (position.column > 2) {
                    //Die Felder bis zur c-Spalte müssen bis auf den Turm leer sein
                    for (column in position.column - 1 downTo 2) {
                        val middlePosition = Position[groundRow, column]
                        val middleFigure = game.getFigureOrNull(middlePosition)
                        if (middleFigure!=null && !middleFigure.canCastle()) {
                            return false
                        }
                    }
                }
                //kommt der Turm auf die d-Linie?
                if (to.column != 3) {
                    val step = (3 - to.column).sign
                    var middlePosition: Position
                    var column = to.column + step
                    while (column != 3) {
                        middlePosition = Position[groundRow, column]
                        val middleFigure = game.getFigureOrNull(middlePosition)
                        if (middleFigure!=null) {
                            if (!middleFigure.canCastle() || middleFigure.isRook()) {
                                return false
                            }
                        }
                        column += step
                    }
                    middlePosition = Position[groundRow, 3]
                    val middleFigure = game.getFigureOrNull(middlePosition)
                    if (middleFigure!=null) {
                        if (!middleFigure.canCastle() || middleFigure.isRook()) {
                            return false
                        }
                    }
                }
                return true
            }
        }
        return false
    }

    fun canNotMoveThereBecauseOfCheck(to: Position, game: SimpleChessBoardInterface, attackLines: AttackLines): Boolean {
        val toFigure = game.getFigureOrNull(to)
        val wantsToCastle = toFigure!=null && toFigure.canCastle()
        if (wantsToCastle) {
            if(attackLines.isCheck) {
                return true
            }
            return isKingAtCheckWhileOrAfterCastling(position, to, game)
        }

        // normal move, castling has been taken care of
        if(attackLines.noCheck) {
            return isKingCheckAt(to, game)
        }
        val directionTo = position.getDirectionTo(to) ?: throw IllegalArgumentException(
                "$to is not next to king's position on: $position"
        )
        for (checkLine in attackLines.checkLines) {
            if(checkLine.keepsKingInCheckIfHeMovesTo(directionTo)) {
                return true
            }
        }
        return isKingCheckAt(to, game)
    }

    private fun isKingAtCheckWhileOrAfterCastling(
            kingPos: Position,
            rookPos: Position,
            game: SimpleChessBoardInterface
    ): Boolean {
        assert(kingPos.row == rookPos.row)

        // king doesn't move to rookPosition so let's determine the real column target
        val updatedToColumn = if (rookPos.column - kingPos.column > 0) 6 else 2

        val columnRange =
                if(updatedToColumn>kingPos.column)
                    (kingPos.column+1)..updatedToColumn
                else updatedToColumn..(kingPos.column-1)

        for (column in columnRange) {
            if (isKingCheckAt(Position[kingPos.row, column], game)) return true
        }
        return false
    }

    private fun isKingCheckAt(to: Position, game: SimpleChessBoardInterface): Boolean {
        val from = position
        val figureTaken = game.move(this, to)
        val isCheck = CheckSearch.isCheck(game, this)
        game.undoMove(this, from, figureTaken)
        return isCheck
    }

    override fun getReachableMoves(game: BasicChessGameInterface, result: MutableList<Move>) = throw UnsupportedOperationException("King doesn't support this method. Use getPossibleKingMoves(..)")
    override fun getPossibleMovesWhileUnboundAndCheck(game: SimpleChessBoardInterface, checkLine: CheckLine, result: MutableList<Move>) = throw UnsupportedOperationException("King doesn't support this method. Use getPossibleKingMoves(..)")
    override fun getPossibleMovesWhileBoundAndNoCheck(game: SimpleChessBoardInterface, boundLine: BoundLine, result: MutableList<Move>) = throw UnsupportedOperationException("King doesn't support this method. Use getPossibleKingMoves(..)")

    override fun getPossibleMoves(game: SimpleChessBoardInterface, result: MutableList<Move>) {
        val attackLines = game.getAttackLines(isWhite)

        if(attackLines.noCheck) {
            for(direction in Direction.values()) {
                position.step(direction)?.let { possibleKingPos->
                    if(isAccessible(game, possibleKingPos) && !isKingCheckAt(possibleKingPos, game)) {
                        result.add(Move[position, possibleKingPos])
                    }
                }
            }

            if(canCastle()) {
                for (column in position.column + 1..7) {
                    val pos = Position[position.row, column]
                    val figure = game.getFigureOrNull(pos)
                    if (figure!=null && figure.canCastle() && isShortCastlingReachable(pos, game)) {
                        if(!isKingAtCheckWhileOrAfterCastling(position, pos, game)) {
                            result.add(Move[position, pos])
                        }
                        break
                    }
                }
                for (column in position.column - 1 downTo 0) {
                    val pos = Position[position.row, column]
                    val figure = game.getFigureOrNull(pos)
                    if (figure!=null && figure.canCastle() && isLongCastlingReachable(pos, game)) {
                        if(!isKingAtCheckWhileOrAfterCastling(position, pos, game)) {
                            result.add(Move[position, pos])
                        }
                        break
                    }
                }
            }
        } else {
            // isSingleCheck || isDoubleCheck
            Direction.values().forEach directionLoop@ {direction ->
                position.step(direction)?.let { possibleKingPos->
                    for(checkLine in attackLines.checkLines) {
                        if(checkLine.keepsKingInCheckIfHeMovesTo(direction)) {
                            return@directionLoop
                        }
                    }
                    if (isAccessible(game, possibleKingPos) && !isKingCheckAt(possibleKingPos, game)) {
                        result.add(Move[position, possibleKingPos])
                    }
                }

            }
        }
    }

    override fun isSelectable(game: SimpleChessBoardInterface): Boolean {
        val minRow = Math.max(position.row - 1, 0)
        val minColumn = Math.max(position.column - 1, 0)
        val maxRow = Math.min(position.row + 1, 7)
        val maxColumn = Math.min(position.column + 1, 7)

        for (row in minRow..maxRow) {
            for (column in minColumn..maxColumn) {
                val checkPosition = Position[row, column]
                if (isMovable(checkPosition, game)) {
                    return true
                }
            }
        }

        if (position.column + 2 < 8) {
            val shortCastling = Position[position.row, position.column + 2]
            if (isMovable(shortCastling, game)) {
                return true
            }
        }

        if (position.column - 2 >= 0) {
            val longCastling = Position[position.row, position.column - 2]
            if (isMovable(longCastling, game)) {
                return true
            }
        }

        return false
    }

    override fun countReachableMoves(game: BasicChessGameInterface): Int {
        var count = 0
        val minRow = Math.max(position.row - 1, 0)
        val minColumn = Math.max(position.column - 1, 0)
        val maxRow = Math.min(position.row + 1, 7)
        val maxColumn = Math.min(position.column + 1, 7)

        for (row in minRow..maxRow) {
            for (column in minColumn..maxColumn) {
                val checkPosition = Position[row, column]
                if (isReachable(checkPosition, game)) {
                    count++
                }
            }
        }

        if (position.column + 2 < 8) {
            val shortCastling = Position[position.row, position.column + 2]
            if (isReachable(shortCastling, game)) {
                count++
            }
        }

        if (position.column - 2 >= 0) {
            val longCastling = Position[position.row, position.column - 2]
            if (isReachable(longCastling, game)) {
                count++
            }
        }

        return count
    }

    fun didCastling() = didCastling

    fun performCastling() {
        didCastling = true
    }

    override fun undoMove(oldPosition: Position) {
        super.undoMove(oldPosition)

        if (stepsTaken == 0) {
            didCastling = false
        }
    }

    override fun toString() = if (didCastling) "${super.toString()}-true" else super.toString()
}
